java-checks/src/main/java/org/sonar/java/checks/xml/struts/ActionNumberCheck.java

```diff
@@ -41,30 +38,25 @@
     defaultValue = "" + DEFAULT_MAXIMUM_NUMBER_FORWARDS)
   public int maximumForwards = DEFAULT_MAXIMUM_NUMBER_FORWARDS;
 
-  private XPathExpression actionsExpression;
-  private XPathExpression forwardsFromActionExpression;
+  private XPathExpression actionsExpression = getXPathExpression("struts-config/action-mappings/action");
+  private XPathExpression forwardsFromActionExpression = getXPathExpression("forward");
 
   @Override
-  public void precompileXPathExpressions(XmlCheckContext context) {
-    actionsExpression = context.compile("struts-config/action-mappings/action");
-    forwardsFromActionExpression = context.compile("forward");
-  }
-
-  @Override
-  public void scanFileWithXPathExpressions(XmlCheckContext context) {
-    for (Node action : context.evaluateOnDocument(actionsExpression)) {
-      Iterable<Node> extraForwards = Iterables.skip(context.evaluate(forwardsFromActionExpression, action), maximumForwards);
-      if (!Iterables.isEmpty(extraForwards)) {
-        List<XmlCheckContext.XmlDocumentLocation> secondaries = new LinkedList<>();
-        for (Node forward : extraForwards) {
-          secondaries.add(new XmlCheckContext.XmlDocumentLocation("Move this forward to another action.", forward));
+  protected void scanFile(XmlFile xmlFile) {
+    evaluateAsList(actionsExpression, xmlFile.getNamespaceUnawareDocument())
+      .forEach(node -> {
+        List<Secondary> secondaries = new ArrayList<>();
```

Comment:
instead of adding element to the list, you can use a map directly to build this list:
```java
List<Secondary> secondaries = evaluateAsList(forwardsFromActionExpression, node).stream()
  .skip(maximumForwards)
  .map(forward -> new Secondary(forward, "Move this forward to another action."))
  .collect(Collectors.toList());
```